
Basic Usage
================================================================================

What you need to run SEISMIC-RNA
--------------------------------------------------------------------------------

- SEISMIC-RNA: to install it, see :doc:`../install`.
- A file of one or more "reference sequences" in FASTA format (for details, see
  :doc:`../formats/data/fasta`).
  This file tells SEISMIC-RNA the underlying sequence of each RNA, to which it
  refers when counting mutations.
  It can contain from one sequence (e.g. if your experiment contained a single
  RNA that you transcribed *in vitro*) up to an entire transcriptome (e.g. if
  you extracted total RNA from cells).
- One or more data files. These can be

  - sequencing reads in FASTQ format (see :doc:`../formats/data/fastq`)
  - alignment maps in SAM/BAM/CRAM format (see :doc:`../formats/data/xam`)
  - reports generated by SEISMIC-RNA (see :doc:`../formats/report/index`)
  - tables generated by SEISMIC-RNA


Run the whole workflow
--------------------------------------------------------------------------------

You can run the entire workflow with one command, ``seismic wf``::

    seismic wf [OPTIONS] FASTA [INPUT_PATH]...

Replace ``FASTA`` with the path to your file of reference sequences (in FASTA
format).
To input FASTQ files, use

- ``-x`` for paired-end reads in separate FASTQ files
- ``-y`` for paired-end reads interleaved in one FASTQ file
- ``-z`` for single-end reads

for example ::

    seismic wf -x sampleA_R1.fq -x sampleA_R2.fq refs.fa

To input alignment maps, reports, or tables, type the paths to the files after
the path to the FASTA file, for example ::

    seismic wf refs.fa out/sampleA/align/ref1.bam

FASTQ files, alignment maps, reports, and tables will enter into the workflow at
these steps:

.. image::
    wf.png
