# Alignment Substitution Analyzer (for nucleotide and aminoacid substitutions)

This project is a Python script that allows researchers to analyze multiple sequence alignments (MSA, as FASTA or Clustalw Files) and generate substitution statistics for nucleotide and protein sequences. This tool is particularly useful for researchers and bioinformaticians studying sequence variation and evolutionary relationships.

DOI/Cite:

Karagol, A., & Karagol, T. (2024). An Evolutionary Statistics Toolkit for Simplified Sequence Analysis on Web with Client-Side Processing. bioRxiv, 2024-08. doi: https://doi.org/10.1101/2024.08.01.606148


### Installation

- Option 1 (RECOMMENDED, for all platforms): You can install the package via pip (if you have Python 3.x on your system), and use the command AlignSubs directly:
```
pip install AlignSubs  
```
```
AlignSubs
```


- Option 2 (for all platforms): You can run the script directly from Google Colab: [AlignSubs.ipynb on Google Colab]
  

- Option 3 (for all platforms): This tool is also available on web: [www.tanerkaragol.com/](https://www.tanerkaragol.com/)

Latest version on PyPI:

[![PyPI version](https://badge.fury.io/py/HGVStoFASTA.svg)](https://badge.fury.io/py/HGVStoFASTA)

### Requirements
If you want to run the script natively on your local computer (Option 1), ensure you have Python 3.x installed on your system. 

If Biopython is not already installed, the script will prompt you to install it automatically. 
If the script fails to automatically install Biopython, you can install it manually using the following steps:

- Open a terminal or command prompt after installing Python.
- Run the following command to install Biopython using pip:
  
   ```
   pip install biopython
   ```


### Questions
If you have any questions or encounter issues, don't hesitate to reach out.

### License
This project is licensed under the  GPL-3.0 License - see the LICENSE file for details.
